queries/bus.py

```diff
@@ -301,81 +257,151 @@ def QBusWord(node: Node, params: QueryStateDict, result: Result) -> None:
-        result.bus_number = int(number)
-    except ValueError:
-        # Handle number words ("sautján")
-        result.bus_number = _BUS_WORDS.get(number, 0)
-        if Settings.DEBUG and result.bus_number == 0:
-            print("Unexpected bus number word: {0}".format(number))
-    except Exception as e:
-        if Settings.DEBUG:
-            print("Unexpected exception: {0}".format(e))
-        raise
+    if "numbers" in result and result.numbers:
+        result.bus_number = result.numbers[0]
 
 
 # End of grammar nonterminal handlers
 
 
-def _filter_func(mm: Iterable[BinEntry]) -> List[BinEntry]:
-    """Filter word meanings when casting bus stop names
-    to cases other than nominative"""
-    # Handle secondary and ternary forms (ÞFFT2, ÞGFET3...)
-    # This is a bit hacky, but necessary for optimal results.
-    # For place names, ÞGFET2 seems often to be a better choice
-    # than ÞGFET, since it has a trailing -i
-    # (for instance 'Skjólvangi' instead of 'Skjólvang')
-    mm2 = [m for m in mm if "ÞGF" in m.mark and "2" in m.mark]
-    if not mm2:
-        # Did not find the preferred ÞGF2, so we go for the
-        # normal form and cut away the secondary and ternary ones
-        mm2 = [m for m in mm if "2" not in m.mark and "3" not in m.mark]
-    return mm2 or list(mm)
+def _compass_directions(stop: str) -> str:
+    """
+    Replace '... N/A/S/V' with '... norður/austur/suður/vestur'
+    for speech synthesis.
+    """
+    if stop.endswith(" N") and straeto.BusStop.named(stop[:-1] + "S"):
+        stop = stop[:-1] + "norður"
+    elif stop.endswith(" A") and straeto.BusStop.named(stop[:-1] + "V"):
+        stop = stop[:-1] + "austur"
+    elif stop.endswith(" S") and straeto.BusStop.named(stop[:-1] + "N"):
+        stop = stop[:-1] + "suður"
+    elif stop.endswith(" V") and straeto.BusStop.named(stop[:-1] + "A"):
+        stop = stop[:-1] + "vestur"
+    return stop
+
+
+_BETTER_NAMES = {
+    "BSÍ": "Umferðarmiðstöðin",
+    "FMOS": "Framhaldsskólinn í Mosfellsbæ",
+    "KEF - Airport": "Keflavíkurflugvöllur",
+    "KR": "Knattspyrnufélag Reykjavíkur",
+    "RÚV": "Útvarpshúsið",
+}
+_ABBREV_RE = re.compile(r"\b[A-ZÁÐÉÍÓÚÝÞÆÖ]+\b")
+
+
+def voicify_stop_name(np: str) -> str:
+    """Fix stop name to better suit speech synthesis."""
+    np = _compass_directions(np)
+    np = _BETTER_NAMES.get(np, np)
+    return _ABBREV_RE.sub(lambda m: gssml(m.group(0), type="spell"), np)
 
 
 @lru_cache(maxsize=None)
-def to_accusative(np: str) -> str:
-    """Return the noun phrase after casting it from nominative to accusative case"""
-    np = straeto.BusStop.voice(np)
-    return queries.to_accusative(np, filter_func=_filter_func)
+def accusative_form(np: str, voice: bool = False) -> str:
+    """
+    Return accusative case of the stop name,
+    optionally expanding abbreviations for speech synthesis.
+    """
+    if voice:
+        np = _compass_directions(np)
+        np = _BETTER_NAMES.get(np, np)
```

Comment:
This logic already exists in voicify_stop_name, abstract it to own fn?

Reply:
Done! Also fixed a small issue with `NounPhrase` not being able to inflect certain stop names with the cardinal direction substituted in for N/A/S/V